auth/auth.go

```diff
@@ -0,0 +1,272 @@
+package auth
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"sync"
+	"time"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/macaroon-bakery.v2/bakery"
+	"gopkg.in/macaroon-bakery.v2/bakery/checkers"
+	"gopkg.in/macaroon-bakery.v2/httpbakery"
+	"gopkg.in/macaroon-bakery.v2/httpbakery/agent"
+	"gopkg.in/macaroon.v2"
+	"gopkg.in/yaml.v2"
+)
+
+const groupScope = "https://www.googleapis.com/auth/admin.directory.group.readonly"
+
+// MacaroonHeadersProvider implements go.temporal.io/sdk/client/internal.HeadersProvider
+// to customize the authorization header when connecting to Temporal instance
+// backed up by Candid auth.
+type MacaroonHeadersProvider struct {
+	bakeryClient *httpbakery.Client
+	authOptions  *MacaroonAuthOptions
+
+	// Ms holds the discharged macaroon in memory while it's still valid (not
+	// expired).
+	ms macaroon.Slice
+	mu sync.Mutex
+}
+
+// GoogleHeadersProvider implements go.temporal.io/sdk/client/internal.HeadersProvider
+// to customize the authorization header when connecting to Temporal instance
+// backed up by Google IAM auth.
+type GoogleHeadersProvider struct {
+	authOptions  *GoogleAuthOptions
+
+	// token holds the access token in memory while it's still valid (not expired).
+	token *oauth2.Token
+}
+
+// AuthOptions represents the necessary data to create a HeadersProvider for
+// authentication and authorization using supported auth providers.
+type AuthOptions struct {
+	Provider string
+	Config any
+}
+
+// MacaroonAuthOptions represents the necessary data to create a HeadersProvider for
+// authentication and authorization using Candid.
+type MacaroonAuthOptions struct {
+	// MacaroonURL represents the HTTP endpoint from which to fetch minted
+	// macaroons.
+	MacaroonURL string `yaml:"macaroon_url"`
+	// AgentKey represents the agent key that will be authenticated with Candid.
+	AgentKey *bakery.KeyPair `yaml:"keys"`
+	// AgentKey represents the agent user name that will be authenticated with Candid.
+	AgentUsername string `yaml:"username"`
+}
+
+// GoogleAuthOptions represents the necessary data to create a HeadersProvider for
+// authentication and authorization using Google IAM.
+type GoogleAuthOptions struct {
+	Type string `yaml:"type"`
```

Comment:
Minor: what about adding godocs for fields here? This will ensure consistency with `MacaroonAuthOptions`.

Also, see my other comment about godoc consistency for methods & functions. I am OK with any approach as long as it's consistent.